s3/s3_transport/include/s3_transport.hpp

```diff
@@ -0,0 +1,1901 @@
+                // the last page might be larger so doing a little trick to handle that case (second term)
+                //  Note:  We bailed early if config_.part_size == 0
+                unsigned long sequence = (file_offset_ / config_.part_size) +
+                    (file_offset_ % config_.part_size == 0 ? 0 : 1) + 1;
+
+                write_callback->sequence = sequence;
+                write_callback->content_length = config_.part_size;
+
+                // estimate the size and resize the etags vector
+                unsigned long  number_of_parts = config_.object_size / config_.part_size;
+                number_of_parts = number_of_parts < sequence ? sequence : number_of_parts;
+
+                // resize the etags vector if necessary
+                int resize_error = shm_obj.atomic_exec([number_of_parts, &shm_obj](auto& data) {
+
+                    if (number_of_parts > data.etags.size()) {
+                        try {
+                            data.etags.resize(number_of_parts, types::shm_char_string("", shm_obj.get_allocator()));
+                        } catch (std::bad_alloc& ba) {
+                            data.last_error_code = error_codes::BAD_ALLOC;
+                            return true;
+                        }
+                    }
+
+                    return false;
+
+                });
+
+                if (resize_error) {
+                    rodsLog(config_.debug_log_level, "Error on reallocation of etags buffer in shared memory.");
+                    return;
+                }
+            }
+
+            write_callback->enable_md5 = config_.enable_md5_flag;
+            write_callback->thread_identifier = get_thread_identifier();
+            write_callback->object_key = object_key_;
+            write_callback->shmem_key = shmem_key_;
+            write_callback->shared_memory_timeout_in_seconds = config_.shared_memory_timeout_in_seconds;
+
+            do {
+
+                if (read_from_cache) {
+                    write_callback->offset = offset;
+                }
+
+                std::stringstream msg;
+
+                msg << "Multipart:  Start part " << static_cast<int>(write_callback->sequence) << ", key \""
+                    << object_key_ << "\", uploadid \"" << upload_id
+                    << "\", len " << static_cast<int>(write_callback->content_length);
+                rodsLog(config_.debug_log_level,  "%s:%d (%s) [[%lu]] %s\n", __FILE__, __LINE__, __FUNCTION__, get_thread_identifier(),
+                        msg.str().c_str() );
+
+                put_props_.md5 = nullptr;
+                //if ( partData.enable_md5 ) {
+                //    // jjames - not sure how to do MD5 piecewise
+                //    putProps->md5 = s3CalcMD5( partData.put_object_data.fd,
+                //    partData.put_object_data.offset,
+                //    partData.put_object_data.content_length, resource_name );
+                //}
```

Comment:
i see the note - just delete this - or keep around for later?

Reply:
deleted.  I can add it back in later.